### PR TITLE
Chore: rework Bazel build system

### PR DESCRIPTION
**Pull request by hcoona**
Totally rewrite the bazel build support.

1. Support latest version of bazel.
2. Build all external dependencies (except for mesalink) with bazel.
3. Support Protobuf 3.19.1 (breaking changes here: https://github.com/protocolbuffers/protobuf/commit/624d29d83306f3ce2c7e092dd44a891e04215e67) 
image

4. Support generate `compile_commands.json` for bazel project.
5. Rework the options flags & move all of them into `//bazel/config`

Leave building with mesalink for future. I think maybe all dependencies should link against mesalink instead of openssl.

### CODE REVIEW

**Comment by chenzhangyi** [comment]
Nice job!

Only one question:  Will the changes break the compilation of users with old versions of bazel, e.g. 0.25

File: src/brpc/protocol.cpp [hunk 1]
```diff
@@ -203,7 +203,7 @@ BUTIL_FORCE_INLINE bool ParsePbFromZeroCopyStreamInlined(
     // According to source code of pb, SetTotalBytesLimit is not a simple set,
     // avoid calling the function when the limit is definitely unreached.
     if (PB_TOTAL_BYETS_LIMITS < FLAGS_max_body_size) {
-        decoder.SetTotalBytesLimit(INT_MAX, -1);
```
Comment (wwbmmm) [inline src/brpc/protocol.cpp:206]: You can check protobuf version and keep compatibility with old version protobuf

#if GOOGLE_PROTOBUF_VERSION >= 3006000
        decoder.SetTotalBytesLimit(INT_MAX);
#else
        decoder.SetTotalBytesLimit(INT_MAX, -1);
#endif

Comment (hcoona) [reply src/brpc/protocol.cpp:206]: fixed

**Comment by hcoona** [comment]
@chenzhangyi In short it would break building with bazel 0.25. Many dependencies relying on a higher version of bazel. I didn't verify each of them. However, `rules_foreign_cc` 0.7.0 requires bazel v4.0.0+ (See https://github.com/bazelbuild/rules_foreign_cc/blob/0.7.0/README.md).

**Comment by hcoona** [comment]
The protobuf arena changes is still breaking. I'll fix it with `GOOGLE_PROTOBUF_VERSION`.

**Comment by hcoona** [comment]
Fixed protobuf compatibility & separate it into another commit.

The failing unit test seems to be unrelated with this change. The only failing unittest is 

image

**Comment by hcoona** [comment]
@wwbmmm PTAL

**Comment by wwbmmm** [comment]
> @wwbmmm PTAL

LGTM

**Comment by hcoona** [comment]
Is there something else to make it merged?

**Comment by lorinlee** [comment]
Hi @hcoona , Thank you so much for your contribution!  But I have a little concern, some projects may depend on bazel 0.2x, like TensorFlow 1.x, and we have seen some teams using bRPC in TensorFlow. So, maybe we should consider the compatibility issues. cc @chenzhangyi @wwbmmm

**Comment by wwbmmm** [comment]
@hcoona Can you extract the changes related to “Support Protobuf 3.19.1” into a separate PR so that this part can be merged faster？

**Comment by hcoona** [comment]
@wwbmmm Sure, split the protobuf support to https://github.com/apache/incubator-brpc/pull/1679

**Comment by hcoona** [comment]
@lorinlee @chenzhangyi In bazel philosophy, the final project owner need to take responsibility for all of its dependencies, AKA. specify how to download & build them in `WORKSPACE` file. For this change, there's nothing special for the `BUILD` files, so it should works fine building a project depends on brpc with an earlier version of Bazel if the project owner could provide necessary indirect dependencies for bRPC. I'll verify it offline, and ping you guys later few days.

**Comment by hcoona** [comment]
Do my best to make it compiles with bazel 0.28.0 (0.27.1 failed). See `example/build_with_old_bazel`.

**Comment by hcoona** [comment]
@chenzhangyi @lorinlee ping~

**Comment by hcoona** [comment]
@chenzhangyi Would you help to take a look at it? The user project depending on bRPC would still compile if they are using bazel v0.28.0 and higher. `example/build_with_old_bazel` verified it.

**Comment by mapx** [comment]
ping

**Comment by zyearn** [comment]
@hcoona Hi, Could you upload the newest version if anything has changed since last time?

**Comment by hcoona** [comment]
@zyearn updated to resolve conflicts & add license headers

**Comment by zyearn** [comment]
Thanks. I will have a look later.

**Comment by 372046933** [comment]
Will it be merged soon?

**Comment by guodongxiaren** [comment]
@hcoona 辛苦解决一下冲突

**Comment by hcoona** [comment]
@guodongxiaren PTAL

**Comment by guodongxiaren** [comment]
@zyearn @wwbmmm please review this pull request and check whether it can be merged.

**Comment by wwbmmm** [comment]
> @zyearn @wwbmmm please review this pull request and check whether it can be merged.

LGTM
@zyearn please see if there is any problem to merge it.

**Comment by guodongxiaren** [comment]
@zyearn @jamesge please review

File: bazel/third_party/glog/0001-improvement-copts.patch [hunk 2]
```diff
@@ -0,0 +1,43 @@
+# Licensed to the Apache Software Foundation (ASF) under one or more
```
Comment (zyearn) [inline bazel/third_party/glog/0001-improvement-copts.patch:1]: What is this file used for? Can we delete it?

File: bazel/third_party/glog/0002-mark-override.patch [hunk 3]
```diff
@@ -0,0 +1,36 @@
+# Licensed to the Apache Software Foundation (ASF) under one or more
```
Comment (zyearn) [inline bazel/third_party/glog/0002-mark-override.patch:1]: What is this file used for? Can we delete it?

File: example/build_with_old_bazel/.bazelrc [hunk 4]
```diff
@@ -0,0 +1,16 @@
+# Licensed to the Apache Software Foundation (ASF) under one or more
```
Comment (zyearn) [inline example/build_with_old_bazel/.bazelrc:1]: Do we need to delete all the files under `build_with_old_bazel`?

Comment (hcoona) [reply example/build_with_old_bazel/.bazelrc:1]: It's to prove we can build brpc with old version of bazel. It's under example folder. We can delete it if we want.

File: bazel/third_party/glog/0002-mark-override.patch (hunk 3 above)
Comment (hcoona) [reply bazel/third_party/glog/0002-mark-override.patch:1]: This solved a noisy compiler warning. I suggest applying it. The final user could use their provided glog. It's harmless.

File: bazel/third_party/glog/0001-improvement-copts.patch (hunk 2 above)
Comment (hcoona) [reply bazel/third_party/glog/0001-improvement-copts.patch:1]: Sure, will do.

**Comment by Huixxi** [comment]
It's a great contribution, and could you please share a related blog or doc for that pr? I think it will be a good learning material and we will post it to our official website, please consider this, thanks a lot.

**Comment by hcoona** [comment]
@Huixxi I didn't realize there was any special work. Please look the Bazel official document & some existing repo driven by Bazel. Please raise to me if there was truly something special with details, I can help to find some document about them.

**Comment by wwbmmm** [comment]
> If you think the example of building with old version of bazel for user is not necessary, I can remove it.

I think the example of building with old version of bazel is useful. No need to remove it.

**Comment by zyearn** [comment]
CI is fixed in the latest version. Could you merge master and submit again?

**Comment by hcoona** [comment]
@zyearn It works after rebase.

**Comment by zyearn** [comment]
LGTM. Thanks for the contribution.

Btw, what do you think is the future work for improving Bazel in brpc?

**Comment by 0x2Adr1** [comment]
Looks like I am coming a few minutes too late after the merge! There is a slight change to be made to make it work with Mac M1:

```
--- BUILD.bazel
+++ BUILD.bazel
@@ -48,7 +48,7 @@ LINKOPTS = [
     "-pthread",
     "-ldl",
 ] + select({
-    "@bazel_tools//tools/osx:darwin": [
+    "@platforms//os:macos": [
         "-framework CoreFoundation",
         "-framework CoreGraphics",
         "-framework CoreData",
@@ -206,7 +206,7 @@ BUTIL_SRCS = [
     "src/butil/recordio.cc",
     "src/butil/popen.cpp",
 ] + select({
-    "@bazel_tools//tools/osx:darwin": [
+    "@platforms//os:macos": [
         "src/butil/time/time_mac.cc",
         "src/butil/mac/scoped_mach_port.cc",
     ],
@@ -322,7 +322,7 @@ cc_library(
         "//bazel/config:brpc_with_glog": ["@com_github_google_glog//:glog"],
         "//conditions:default": [],
     }) + select({
-        "@bazel_tools//tools/osx:darwin": [":macos_lib"],
+        "@platforms//os:macos": [":macos_lib"],
         "//conditions:default": [],
     }),
 )
 ```

**Comment by zyearn** [comment]
@0x2Adr1 Could you make a PR separately?

**Comment by hcoona** [comment]
> LGTM. Thanks for the contribution.
> 
> Btw, what do you think is the future work for improving Bazel in brpc?

It's good enough now. I don't think we need to adjust it unless we plan to change the project structure or Bazel bring a big change.

**Comment by guodongxiaren** [comment]
@hcoona MacOs上使用bazel似乎还有问题。烦请试试

**Comment by hcoona** [comment]
@guodongxiaren I've no macOS environment. Please help to post the error message & I'll try to investigate it. And please notice that @0x2Adr1 raise some additional fixes for macOS, please also try it.

**Comment by Huixxi** [comment]
It's better to update the [getting_start document](https://github.com/apache/incubator-brpc/blob/master/docs/cn/getting_started.md)  (cn/en) with a new "Compile brpc with bazel" part under each platform, but I'm not familiar with bazel...

**Pull request merged**